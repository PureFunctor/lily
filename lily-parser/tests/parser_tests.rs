use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
