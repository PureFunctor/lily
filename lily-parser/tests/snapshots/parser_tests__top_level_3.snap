---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 20,
                end: 99,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 30,
                        end: 99,
                        kind: IfThenElse(
                            Expression {
                                begin: 33,
                                end: 51,
                                kind: IfThenElse(
                                    Expression {
                                        begin: 36,
                                        end: 37,
                                        kind: Variable(
                                            "a",
                                        ),
                                    },
                                    Expression {
                                        begin: 43,
                                        end: 44,
                                        kind: Variable(
                                            "b",
                                        ),
                                    },
                                    Expression {
                                        begin: 50,
                                        end: 51,
                                        kind: Variable(
                                            "c",
                                        ),
                                    },
                                ),
                            },
                            Expression {
                                begin: 57,
                                end: 75,
                                kind: IfThenElse(
                                    Expression {
                                        begin: 60,
                                        end: 61,
                                        kind: Variable(
                                            "d",
                                        ),
                                    },
                                    Expression {
                                        begin: 67,
                                        end: 68,
                                        kind: Variable(
                                            "e",
                                        ),
                                    },
                                    Expression {
                                        begin: 74,
                                        end: 75,
                                        kind: Variable(
                                            "f",
                                        ),
                                    },
                                ),
                            },
                            Expression {
                                begin: 81,
                                end: 99,
                                kind: IfThenElse(
                                    Expression {
                                        begin: 84,
                                        end: 85,
                                        kind: Variable(
                                            "g",
                                        ),
                                    },
                                    Expression {
                                        begin: 91,
                                        end: 92,
                                        kind: Variable(
                                            "h",
                                        ),
                                    },
                                    Expression {
                                        begin: 98,
                                        end: 99,
                                        kind: Variable(
                                            "i",
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
            },
        ],
    },
)
